File: test/webui/orchestrate.test.js
'use strict';

// The Orchestration switch in the settings window.
//
// A second switch beside Smart Agent, and the reason there are two is the
// reason these tests exist: the two are different sizes. Smart Agent lets
// the model hand one question to a specialist; this lets it commit to a
// shape and spend up to thirty-two agent turns on it. So they have to move
// independently, and the panel has to be able to show one on and the other
// off without asking the daemon twice.

const test = require('node:test');
const assert = require('node:assert/strict');

const { load } = require('./harness');

const SETTINGS = {
  auto_compact_enabled: true, show_tps: true, auto_delegate: false,
  auto_delegate_agent: '', auto_delegate_match: [],
  smart_agent: false, smart_agent_roster: ['explore', 'oracle'],
  orchestrate: false,
  skip_permissions: false, permission_rules: {}, can_edit_permissions: true,
};

async function openSettings(app) {
  app.el('settings-btn').click();
  await app.settle();
}

test('the switch is off, and says what that means', async () => {
  const app = await load({ routes: { 'GET /api/settings': SETTINGS } });
  await openSettings(app);

  assert.equal(app.el('orchestrate-checkbox').checked, false);
  assert.match(app.el('orchestrate-note').textContent, /^Off\./);
  assert.equal(app.callsTo('POST', '/api/settings/orchestrate').length, 0,
    'opening the panel changed a setting');
});

// On, and the note carries the ceilings. They are the part a person needs
// before agreeing to it: a run is the most expensive single thing
// localcode can be asked to do.
test('on, it names what a run can cost', async () => {
  const app = await load({
    routes: { 'GET /api/settings': { ...SETTINGS, smart_agent: true, orchestrate: true } },
  });
  await openSettings(app);

  assert.equal(app.el('orchestrate-checkbox').checked, true);
  const note = app.el('orchestrate-note').textContent;
  assert.match(note, /8 stages and 32 agent turns/);
  assert.match(note, /asks first/);
});

// A plan needs somewhere to delegate its stages to. Turning this on with
// Smart Agent off is legal and inert, and saying so beats letting it read
// as a change that took effect.
test('on with nobody to delegate to, it says so instead of pretending', async () => {
  const app = await load({
    routes: { 'GET /api/settings': { ...SETTINGS, smart_agent: false, orchestrate: true } },
  });
  await openSettings(app);

  assert.match(app.el('orchestrate-note').textContent, /nobody to delegate to/);
  assert.match(app.el('orchestrate-note').textContent, /Turn on Smart Agent/);
});

test('ticking it tells the daemon', async () => {
  const app = await load({
    routes: {
      'GET /api/settings': SETTINGS,
      'POST /api/settings/orchestrate': { orchestrate: true, applied: true, persisted: true },
    },
  });
  await openSettings(app);

  app.el('orchestrate-checkbox').checked = true;
  app.el('orchestrate-checkbox').fire('change');
  await app.settle();

  const calls = app.callsTo('POST', '/api/settings/orchestrate');
  assert.equal(calls.length, 1);
  assert.deepEqual(calls[0].body, { enabled: true });
  assert.equal(app.el('orchestrate-checkbox').checked, true);
});

// Applied and saved are different questions, and the daemon answers both.
// A change that reached the running daemon but not config.json is still a
// change: the box has to keep showing it, with the warning beside it.
test('applied but not saved keeps the box and warns', async () => {
  const app = await load({
    routes: {
      'GET /api/settings': SETTINGS,
      'POST /api/settings/orchestrate': {
        orchestrate: true, applied: true, persisted: false, error: 'permission denied',
      },
    },
  });
  await openSettings(app);

  app.el('orchestrate-checkbox').checked = true;
  app.el('orchestrate-checkbox').fire('change');
  await app.settle();

  assert.equal(app.el('orchestrate-checkbox').checked, true,
    'an unsaved change was shown as a refused one, so the box now denies the state the daemon is in');
  assert.match(app.el('orchestrate-warn').textContent, /not saved to config\.json/);
  assert.equal(app.el('orchestrate-warn').hidden, false);
});

// Nothing was applied. Put the box back.
test('a refused change puts the box back', async () => {
  const app = await load({
    routes: {
      'GET /api/settings': SETTINGS,
      'POST /api/settings/orchestrate': { status: 500, body: 'nope' },
    },
  });
  await openSettings(app);

  app.el('orchestrate-checkbox').checked = true;
  app.el('orchestrate-checkbox').fire('change');
  await app.settle();

  assert.equal(app.el('orchestrate-checkbox').checked, false);
  assert.match(app.el('orchestrate-note').textContent, /^Not changed:/);
});

// The two switches are independent, which is the whole argument for there
// being two of them.
test('the two switches are drawn from one payload and move apart', async () => {
  const app = await load({
    routes: { 'GET /api/settings': { ...SETTINGS, smart_agent: true, orchestrate: false } },
  });
  await openSettings(app);

  assert.equal(app.el('smart-agent-checkbox').checked, true);
  assert.equal(app.el('orchestrate-checkbox').checked, false);
});

File: internal/daemon/orchestrate_test.go
package daemon

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// The endpoint behind the settings panel's Orchestration switch.
//
// Its own switch rather than a line inside Smart Agent, because the two
// are different sizes: Smart Agent lets the model hand one question to a
// specialist, this lets it commit to a shape and spend up to thirty-two
// agent turns on it. So the two have to move independently, and the panel
// has to be able to show one on and the other off.

func TestDaemonSetOrchestrate(t *testing.T) {
	model := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}))
	defer model.Close()

	d := newTestDaemon(t, model.URL)
	cfgPath := filepath.Join(t.TempDir(), "config.json")
	if err := os.WriteFile(cfgPath, []byte(`{"show_tps":false}`), 0o644); err != nil {
		t.Fatal(err)
	}
	d.Broker.ConfigPath = cfgPath

	httpSrv := httptest.NewServer(d.Handler())
	defer httpSrv.Close()

	// A run is the most expensive single thing localcode can be asked to
	// do. A build that shipped with this on would be spending it for
	// everyone who never asked.
	if d.Loop.OrchestrateEnabled() {
		t.Fatal("orchestration started on, so this test would prove nothing")
	}

	post := func(enabled bool) {
		t.Helper()
		body := strings.NewReader(fmt.Sprintf(`{"enabled":%t}`, enabled))
		resp, err := http.Post(httpSrv.URL+"/api/settings/orchestrate", "application/json", body)
		if err != nil {
			t.Fatalf("POST orchestrate: %v", err)
		}
		defer resp.Body.Close()
		if resp.StatusCode != http.StatusOK {
			t.Fatalf("POST orchestrate status = %d, want 200", resp.StatusCode)
		}
		var got struct {
			Orchestrate bool `json:"orchestrate"`
			Applied     bool `json:"applied"`
			Persisted   bool `json:"persisted"`
		}
		if err := json.NewDecoder(resp.Body).Decode(&got); err != nil {
			t.Fatalf("decode: %v", err)
		}
		if got.Orchestrate != enabled || !got.Applied || !got.Persisted {
			t.Fatalf("response = %+v, want orchestrate=%t applied and persisted", got, enabled)
		}
	}

	post(true)
	if !d.Loop.OrchestrateEnabled() {
		t.Error("the running daemon did not change")
	}
	// Saved as a top-level boolean. The shape matters: a nested object
	// written here would be a config.json that no later load can decode.
	saved, err := os.ReadFile(cfgPath)
	if err != nil {
		t.Fatal(err)
	}
	var raw map[string]any
	if err := json.Unmarshal(saved, &raw); err != nil {
		t.Fatalf("the config we wrote no longer parses: %v\n%s", err, saved)
	}
	if raw["orchestrate"] != true {
		t.Errorf("config.json has orchestrate=%v, want true:\n%s", raw["orchestrate"], saved)
	}
	if raw["show_tps"] != false {
		t.Error("writing the switch dropped a key that was already in the file")
	}

	post(false)
	if d.Loop.OrchestrateEnabled() {
		t.Error("turning it off did not change the running daemon")
	}

	// And the two switches are genuinely separate, which is the whole
	// argument for there being two.
	d.Loop.SetSmartAgentEnabled(true)
	post(false)
	if !d.Loop.SmartAgentEnabled() {
		t.Error("changing orchestration turned Smart Agent off")
	}
}

// The panel reads both switches from one GET, so a client can show one on
// and the other off without asking twice.
func TestSettingsReportsBothSwitches(t *testing.T) {
	model := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}))
	defer model.Close()
	d := newTestDaemon(t, model.URL)
	d.Loop.SetSmartAgentEnabled(true)
	d.Loop.SetOrchestrateEnabled(false)

	httpSrv := httptest.NewServer(d.Handler())
	defer httpSrv.Close()

	resp, err := http.Get(httpSrv.URL + "/api/settings")
	if err != nil {
		t.Fatalf("GET settings: %v", err)
	}
	defer resp.Body.Close()
	var got map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&got); err != nil {
		t.Fatalf("decode: %v", err)
	}
	if got["smart_agent"] != true {
		t.Errorf("smart_agent = %v, want true", got["smart_agent"])
	}
	if _, ok := got["orchestrate"]; !ok {
		t.Fatal("the settings payload has no orchestrate key, so the panel cannot draw the switch")
	}
	if got["orchestrate"] != false {
		t.Errorf("orchestrate = %v, want false", got["orchestrate"])
	}
}

// A body that is not what the endpoint takes is a 400, not a silent
// change: the switch decides what a request may spend.
func TestSetOrchestrateRefusesAJunkBody(t *testing.T) {
	model := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {}))
	defer model.Close()
	d := newTestDaemon(t, model.URL)
	httpSrv := httptest.NewServer(d.Handler())
	defer httpSrv.Close()

	resp, err := http.Post(httpSrv.URL+"/api/settings/orchestrate", "application/json",
		strings.NewReader(`not json`))
	if err != nil {
		t.Fatalf("POST: %v", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusBadRequest {
		t.Errorf("status = %d, want 400", resp.StatusCode)
	}
	if d.Loop.OrchestrateEnabled() {
		t.Error("a rejected request changed the switch anyway")
	}
}
